### PR TITLE
Expose extended metadata to Rust binding

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -118,6 +146,35 @@ impl Drop for Model {
...
+use std::fmt;
+impl fmt::Display for Metadata {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		unsafe {
+			let num_items = (*self.metadata).num_items;
+			let vec_metadata = &[(*self.metadata).items];
+			let r = {
+				let mut v = Vec::new();
+				for item in 0..num_items {
+					let ptr = (*vec_metadata[item as usize]).character;
```
Comment (lissyx) [inline src/lib.rs]: This is not working but I'm not able to find proper docs on how this should be dealt with, accessing the `items` field of `struct Metadata`.

Comment (est31) [reply src/lib.rs]: In general, to index C arrays, the [std::slice::from_raw_parts](https://doc.rust-lang.org/stable/std/slice/fn.from_raw_parts.html) function is used. I'm not really sure whether the Display implementation should deal with this though, see my other comment.

**Review by est31 — CHANGES_REQUESTED** [review]
Hi, thanks for opening the PR.

Could you also update the sys/deepspeech submodule? That's where the sys crate gets the header from, so such an update is always needed.

As for the API, I would suggest a more Rust-y one than "just" having a Display implementation. I suggest you do the following:

* Add a `#[repr(transparent)]` in front of `MetadataItem` and change the item inside to `metadata_item :ds::MetadataItem` (removing the pointer)

* Add a `pub fn items(&self) -> &[MetadataItem]` function to the `Metadata` struct that a) converts the `* const ds::MetadataItem` pointer to `* const MetadataItem` using `as` casting and b) uses `from_raw_parts` to convert the pointer to a slice ` &[MetadataItem]` that it returns

* Add a function `pub fn character(&self) -> Result<str, ...>` to the `MetadataItem` struct

* Add functions `pub fn timestep(&self) -> i16` and `pub fn start_time(&self) -> f32` to the `MetadataItem` struct

* Add a function `pub fn probability(&self) -> f64` to the `Metadata` struct

**Comment by est31** [comment]
The idea for `#[repr(transparent)]` came from @nox, good catch thank you!

**Comment by lissyx** [comment]
> I would suggest a more Rust-y one than "just" having a Display implementation

This is just for debug purpose, no way intended to be fina

**Comment by lissyx** [comment]
@est31 That should now look closer to what we did on https://github.com/mozilla/DeepSpeech/pull/2022, and implements all your suggestions. I'm leaving `fm::Display` in place, just in case.

File: examples/client_extended.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		let conv = Converter::from_hz_to_hz(
+			from_iter(reader.samples::<i16>().map(|s| [s.unwrap()])),
+			interpolator,
+			desc.sample_rate() as f64,
+			SAMPLE_RATE as f64);
+		conv.until_exhausted().map(|v| v[0]).collect()
+	};
+
+	// Run the speech to text algorithm
+	let result = metadata_to_string(m.speech_to_text_with_metadata(&audio_buf, SAMPLE_RATE).unwrap());
```
Comment (est31) [inline examples/client_extended.rs:82]: You can use the Display implementation instead. Just remove the function and do `result = m.speech_to_text_with_metadata(&audio_buf, SAMPLE_RATE).unwrap();`.

File: src/lib.rs [hunk 3]
```diff
@@ -155,14 +240,27 @@ impl Stream {
...
+			ds::DS_FreeString(ptr);
 			v
 		};
 		// Don't run the destructor for self,
 		// as DS_FinishStream already does it for us
 		forget(self);
 		String::from_utf8(r)
 	}
+
+    /// Deallocates the stream and returns the extended metadata
```
Comment (est31) [inline src/lib.rs]: Please use tabs here

File: src/lib.rs [hunk 4]
```diff
@@ -155,14 +240,27 @@ impl Stream {
...
+		let ptr = unsafe {
+			ds::DS_FinishStreamWithMetadata(self.stream)
+		};
+		// Don't run the destructor for self,
+		// as DS_FinishStream already does it for us
+		forget(self);
+		Ok(Metadata {
+			metadata : ptr
+		})
+    }
```
Comment (est31) [inline src/lib.rs]: Please use tabs here

File: src/lib.rs [hunk 5]
```diff
@@ -118,6 +146,63 @@ impl Drop for Model {
...
+		self.metadata_item.start_time
+	}
+}
+
+use std::fmt;
+impl fmt::Display for Metadata {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		let mut v = Vec::new();
+		for item in 0..self.num_items() {
+			v.extend_from_slice(self.items()[item as usize].character().unwrap().as_bytes());
```
Comment (est31) [inline src/lib.rs]: This isn't idiomatic Rust. You convert from a `str` back to a `&[u8]` and down below you convert back to a `String` again. Instead, make `v` a `String` and append the `str`'s to it via `+=`. Calling `write!` inside the loop would work as well, tbh I don't know which of the two is more idiomatic.

File: src/lib.rs [hunk 6]
```diff
@@ -118,6 +146,63 @@ impl Drop for Model {
...
+	pub fn timestep(&self) -> i32 {
+		self.metadata_item.timestep
+	}
+
+	pub fn start_time(&self) -> f32 {
+		self.metadata_item.start_time
+	}
+}
+
+use std::fmt;
```
Comment (est31) [inline src/lib.rs]: Could you put this use to the top of the file?

File: src/lib.rs [hunk 7]
```diff
@@ -118,6 +146,63 @@ impl Drop for Model {
...
+		unsafe {
+			ds::DS_FreeMetadata(self.metadata);
+		}
+	}
+}
+
+impl Metadata {
+	pub fn items(&self) -> &[MetadataItem] {
+		unsafe {
+			let ptr   = (*self.metadata).items as * const MetadataItem;
```
Comment (est31) [inline src/lib.rs]: You've got three spaces here.

File: src/lib.rs [hunk 8]
```diff
@@ -10,13 +10,22 @@ use std::path::Path;
...
 	model :* mut ds::ModelState,
 }
 
+pub struct Metadata {
+	metadata :* mut ds::Metadata,
+}
+
+#[repr(transparent)]
+pub struct MetadataItem {
+	metadata_item : ds::MetadataItem,
```
Comment (est31) [inline src/lib.rs]: Could you do `metadata_item :ds::MetadataItem`? My code style has `a :b` when b is a type, and `a : b` when b is a value that `a` is assigned to.

**Review by est31 — CHANGES_REQUESTED** [review]
Overall it's much better now. Ready for merge once the things I pointed out are fixed.

File: examples/client_extended.rs (hunk 2 above)
Comment (lissyx) [reply examples/client_extended.rs:82]: Well, in all other bindings we have this `metadata_to_string`, so I thought it would be better to be consistent.

Comment (est31) [reply examples/client_extended.rs:82]: We agreed on IRC to keep this

**Review by est31 — APPROVED** [review]

**Pull request merged**